source/common/callback/manager.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <list>
+
+#include "common/callback/callback.h"
+
+namespace Envoy {
+namespace Common {
+namespace Callback {
+
+/**
+ * Callback::Manager is a fairly typical implementation of the "Observer" pattern
+ * (https://en.wikipedia.org/wiki/Observer_pattern), made safe by using Callback::Caller. Any number
+ * of Callers may be added to a Manager. Resetting or destroying a Caller's corresponding Receiver
+ * will remove it from the Manager's list when it is invoked.
+ *
+ * Manager is actually a type alias (below) to this class template, ManagerT, which is parameterized
+ * on an additional type C, representing the caller type. This will typically be Caller<Args...>,
+ * but can be anything that behaves like a function. See examples in manager_test.cc (where the
+ * caller type is a mock), and Init::ManagerImpl (where the caller does some extra logging).
+ */
+template <typename C, typename... Args> class ManagerT {
+public:
+  /**
+   * Adds a Caller to the callback manager, such that its corresponding Receiver will be invoked
+   * by subsequent calls to call() as long as it remains available.
+   * @param caller the caller to add to the callback manager
+   */
+  ManagerT& add(C caller) {
+    callers_.push_back(std::move(caller));
+    return *this;
+  }
+
+  /**
+   * Invokes all callers previously provided to add(). Any corresponding receivers that are no
+   * longer available after invocation (which may have happened as a side-effect of invoking them)
+   * will be removed from the callback manager.
+   * @param args the arguments, if any, to pass to all callers.
+   */
+  void operator()(Args... args) {
```

Comment:
Do we need to template this? Are args generally super useful to users? Previously we just had `()`.

Reply:
Yes, it's used here. See e.g. `TargetCaller` / `TargetReceiver`. I'm also sort of hoping that this will replace the existing callback mechanism in `source/common/common`, which accepts args.

Reply:
Can you give some examples of where we really need this and can't just get away with passing in a `std::function` that captures its arguments? I think one of the main opportunities to simplify is to reduce the amount of templating here.

Reply:
I think the example I wrote up in the long comment block in callback.h is a pretty good illustration. Suppose you want to register a bunch of callback functions that consume an event, and pass them the event later.